server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java

```diff
@@ -21,48 +21,147 @@
+import java.util.TimeZone;
 import java.util.TreeMap;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Collectors;
+import lombok.Getter;
 import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.hadoop.conf.Configuration;
 
+@Slf4j
 public class MemoryTableImpl implements MemoryTable {
 
+    private static final String LOG_ID_KEY = "logId";
+    private static final String LAST_UPDATE_TIME_KEY = "lastUpdateTime";
+
     private final String tableName;
 
     private final WalManager walManager;
+    private final StorageAccessService storageAccessService;
+    private final String dataPathPrefix;
+    private final SimpleDateFormat dataPathSuffixFormat = new SimpleDateFormat("yyMMddHHmmss.SSS");
+    private final ParquetConfig parquetConfig;
 
     private TableSchema schema = null;
 
+    @Getter
+    private long firstWalLogId = -1;
+
+    private long lastWalLogId = -1;
+
+    @Getter
+    private long lastUpdateTime = 0;
+
     private final TreeMap<Object, Map<String, Object>> recordMap = new TreeMap<>();
 
     private final Lock lock = new ReentrantLock();
 
-    public MemoryTableImpl(String tableName, WalManager walManager) {
+    public MemoryTableImpl(String tableName,
+            WalManager walManager,
+            StorageAccessService storageAccessService,
+            String dataRootPath,
+            ParquetConfig parquetConfig) {
         this.tableName = tableName;
         this.walManager = walManager;
+        this.storageAccessService = storageAccessService;
+        this.parquetConfig = parquetConfig;
+        var path = dataRootPath;
+        if (!path.isEmpty() && !path.endsWith("/")) {
+            path += "/";
+        }
+        this.dataPathPrefix = path + "snapshots/" + tableName + "_._v0_";
+        this.dataPathSuffixFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
+        this.load();
+    }
+
+    private void load() {
+        try {
+            var path = this.storageAccessService.list(dataPathPrefix).max(Comparator.naturalOrder()).orElse(null);
+            if (path == null) {
+                return;
+            }
+            var conf = new Configuration();
+            try (var reader = new SwParquetReaderBuilder(this.storageAccessService, path).withConf(conf).build()) {
+                for (; ; ) {
+                    var record = reader.read();
+                    if (record == null) {
+                        break;
+                    }
+                    if (this.schema == null) {
+                        this.schema = TableSchema.fromJsonString(conf.get(SwReadSupport.SCHEMA_KEY));
+                        var metadata = new ObjectMapper().readValue(conf.get(SwReadSupport.META_DATA_KEY), Map.class);
+                        this.lastWalLogId = Long.parseLong((String) metadata.get(LOG_ID_KEY));
+                        this.lastUpdateTime = Long.parseLong((String) metadata.get(LAST_UPDATE_TIME_KEY));
```

Comment:
Would it be better if `MemoryTableImpl` does not have to care about details in `SwReadSupport`?
Maybe pass `this` to the reader as schemaAndMetaAccepter interface impl?

Reply:
I guess it is a little overdesign. I would like to assign those members explicitly rather than in another class through the setter interface. The latter hurts code readability.

Reply:
ok